src/workerd/api/sockets.c++

```diff
@@ -47,52 +47,68 @@ jsg::Ref<Socket> connectImpl(
   auto sysStreams = newSystemMultiStream(kj::addRef(*stream), StreamEncoding::IDENTITY, context);
+  auto readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable));
+  auto writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable));
+
+  auto closeFulfiller = IoContext::current().addObject(
+      kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()));
+  auto closeFulfillerFork = closeFulfiller->promise.fork();
+
+  // Attach a callback to close the readable/writable streams when the socket is closed (either
+  // explicitly or implicitly).
+  context.awaitIo(js, closeFulfillerFork.addBranch(), [readable = readable.addRef(),
+      writable = writable.addRef()](jsg::Lock& js) mutable {
+    readable->cancel(js, nullptr);
+    writable->abort(js, nullptr);
+  });
 
   return {
-    .readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable)),
-    .writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable)),
-    .closeFulfiller = IoContext::current().addObject(
-        kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()))
+    .readable = kj::mv(readable),
+    .writable = kj::mv(writable),
+    .closeFulfiller = kj::mv(closeFulfiller),
+    .closeFulfillerFork = kj::mv(closeFulfillerFork)
   };
 }
 
-Socket::Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
-    Socket(initialiseSocket(kj::mv(connectionPromise))) {};
+Socket::Socket(jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
+    Socket(initialiseSocket(js, kj::mv(connectionPromise), [this]() { close(); })) {};
 
-jsg::Promise<void> Socket::close(jsg::Lock& js) {
+void Socket::close() {
   if (!closeFulfiller->fulfiller->isWaiting()) {
-    return js.resolvedPromise();
+    return;
   }
 
-  auto result = js.resolvedPromise();
-  result = readable->cancel(js, nullptr);
-  result = writable->abort(js, nullptr);
+  // Readable and writable streams are closed inside callback set up in `initialiseSocket`.
   closeFulfiller->fulfiller->fulfill();
-  return result;
 }
 
-PipelinedAsyncIoStream::PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner) :
-    inner(kj::mv(inner)) {};
+PipelinedAsyncIoStream::PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner,
+    std::function<void()> onClose) :
+    inner(kj::mv(inner)), onClose(onClose) {};
 
 void PipelinedAsyncIoStream::shutdownWrite() {
-  thenOrRunNow<void>([](kj::Own<kj::AsyncIoStream>* stream) {
+  thenOrRunNow<void>([this](kj::Own<kj::AsyncIoStream>* stream) {
     (*stream)->shutdownWrite();
+    onClose();
     return kj::READY_NOW;
   }).detach([this](kj::Exception&& exception) mutable {
     error = kj::mv(exception);
   });
 }
 
 void PipelinedAsyncIoStream::abortRead() {
-  thenOrRunNow<void>([](kj::Own<kj::AsyncIoStream>* stream) {
+  thenOrRunNow<void>([this](kj::Own<kj::AsyncIoStream>* stream) {
```

Comment:
Passing in a pointer to the kj::Own is a bit awkward. Could these use `kj::AsyncIoStream&` instead?

Reply:
Since this PR doesn't introduce these I hope you won't mind if I clean this up in a separate PR.